File: php/lambda.php
<?php
function toInteger($proc) {
  return $proc(function($n) { return $n + 1; })(0);
}

function toBoolean($proc) {
  return $proc(true)(false);
}

$toArray = function($proc) use (&$isEmpty, &$first, &$rest) {
  $array = [];

  while (!toBoolean($isEmpty($proc))) {
    $array[] = $first($proc);
    $proc = $rest($proc);
  }

  return $array;
};

function toChar($c) {
  return "0123456789BFiuz"[toInteger($c)];
}

$toString = function($s) use (&$toArray) {
  return implode('', array_map('toChar', $toArray($s)));
};

$zero  = function($p) { return function($x) { return $x; }; };
$one   = function($p) { return function($x) use (&$p) { return $p($x); }; };
$two   = function($p) { return function($x) use (&$p) { return $p($p($x)); }; };
$three = function($p) { return function($x) use (&$p) { return $p($p($p($x))); }; };

$five    = function($p) { return function($x) use (&$p) { return $p($p($p($p($p($x))))); }; };
$ten     = function($p) { return function($x) use (&$p) { return $p($p($p($p($p($p($p($p($p($p($x)))))))))); }; };
$fifteen = function($p) { return function($x) use (&$p) { return $p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($x))))))))))))))); }; };
$hundred = function($p) { return function($x) use (&$p) { return $p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($p($x)))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))); }; };

$true  = function($x) { return function($y) use (&$x) { return $x; }; };
$false = function($x) { return function($y) { return $y; }; };

$if = function($b) { return $b; };

$isZero = function($n) use (&$true, &$false) { return $n(function($x) use (&$false) { return $false; })($true); };

$increment = function($n) { return function($p) use (&$n) { return function($x) use (&$n, &$p) { return $p($n($p)($x)); }; }; };
$decrement = function($n) { return function($f) use (&$n) { return function($x) use (&$f, &$n) {
  return $n(function($g) use (&$f) { return function($h) use (&$f, &$g) { return $h($g($f)); }; })(function($u) use (&$x) { return $x; })(function($u) { return $u; });
}; }; };

$add      = function($m) use (&$increment)       { return function($n) use (&$increment, &$m)       { return $n($increment)($m); }; };
$subtract = function($m) use (&$decrement)       { return function($n) use (&$decrement, &$m)       { return $n($decrement)($m); }; };
$multiply = function($m) use (&$add, &$zero)     { return function($n) use (&$add, &$zero, &$m)     { return $n($add($m))($zero); }; };
$power    = function($m) use (&$multiply, &$one) { return function($n) use (&$multiply, &$one, &$m) { return $n($multiply($m))($one); }; };

$isLessOrEqual = function($m) use (&$isZero, &$subtract) { return function($n) use (&$isZero, &$subtract, &$m) {
  return $isZero($subtract($m)($n));
}; };

$z = function($f) {
  return (function($x) use (&$f) {
    return $f(function($y) use (&$x) { return $x($x)($y); });
  })(function($x) use (&$f) {
    return $f(function($y) use (&$x) { return $x($x)($y); });
  });
};

$mod = $z(function($f) use (&$if, &$isLessOrEqual, &$subtract) { return function($m) use (&$if, &$isLessOrEqual, &$subtract, &$f) { return function($n) use (&$if, &$isLessOrEqual, &$subtract, &$f, &$m) {
  return $if($isLessOrEqual($n)($m))(
    function($x) use (&$subtract, $f, &$m, &$n) { return $f($subtract($m)($n))($n)($x); }
  )(
    $m
  );
}; }; });

$div = $z(function($f) use (&$if, &$isLessOrEqual, &$increment, &$subtract, &$zero) { return function($m) use (&$if, &$isLessOrEqual, &$increment, &$subtract, &$zero, &$f) { return function($n) use (&$if, &$isLessOrEqual, &$increment, &$subtract, &$zero, &$f, &$m) {
  return $if($isLessOrEqual($n)($m))(
    function($x) use (&$increment, &$subtract, &$f, &$m, &$n) { return $increment($f($subtract($m)($n))($n))($x); }
  )(
    $zero
  );
}; }; });

$pair  = function($x) { return function($y) use (&$x) { return function($f) use (&$x, &$y) { return $f($x)($y); }; }; };
$left  = function($p) { return $p(function($x) { return function($y) use (&$x) { return $x; }; }); };
$right = function($p) { return $p(function($x) { return function($y) { return $y; }; }); };

$empty   = $pair($true)($true);
$unshift = function($l) use (&$pair, &$false) { return function($x) use (&$pair, &$false, &$l) { return $pair($false)($pair($x)($l)); }; };
$isEmpty = $left;
$first   = function($l) use (&$left, &$right) { return $left($right($l)); };
$rest    = function($l) use (&$right) { return $right($right($l)); };

$range = $z(function($f) use (&$if, &$isLessOrEqual, &$unshift, &$increment, &$empty) { return function($m) use (&$if, &$isLessOrEqual, &$unshift, &$increment, &$empty, &$f) { return function($n) use (&$if, &$isLessOrEqual, &$unshift, &$increment, &$empty, &$f, &$m) {
  return $if($isLessOrEqual($m)($n))(
    function($x) use (&$unshift, &$increment, &$f, &$m, &$n) { return $unshift($f($increment($m))($n))($m)($x); }
  )(
    $empty
  );
}; }; });

$fold = $z(function($f) use (&$if, &$isEmpty, &$rest, &$first) { return function($l) use (&$if, &$isEmpty, &$rest, &$first, &$f) { return function($x) use (&$if, &$isEmpty, &$rest, &$first, &$f, &$l) { return function($g) use (&$if, &$isEmpty, &$rest, &$first, &$f, &$l, &$x) {
  return $if($isEmpty($l))(
    $x
  )(
    function($y) use (&$rest, &$first, &$f, &$l, &$x, &$g) { return $g($f($rest($l))($x)($g))($first($l))($y); }
  );
}; }; }; });

$map = function($k) use (&$fold, &$empty, &$unshift) { return function($f) use (&$fold, &$empty, &$unshift, &$k) {
  return $fold($k)($empty)(
    function($l) use (&$unshift, &$f) { return function($x) use (&$unshift, &$f, &$l) { return $unshift($l)($f($x)); }; }
  );
}; };

$push = function($l) use (&$fold, &$unshift, &$empty) { return function($x) use (&$fold, &$unshift, &$empty, &$l) {
  return $fold($l)($unshift($empty)($x))($unshift);
}; };

$b   = $ten;
$f   = $increment($b);
$i   = $increment($f);
$u   = $increment($i);
$zed = $increment($u);

$fizz     = $unshift($unshift($unshift($unshift($empty)($zed))($zed))($i))($f);
$buzz     = $unshift($unshift($unshift($unshift($empty)($zed))($zed))($u))($b);
$fizzbuzz = $unshift($unshift($unshift($unshift($buzz)($zed))($zed))($i))($f);

$toDigits = $z(function($f) use (&$push, &$if, &$isLessOrEqual, &$decrement, &$ten, &$empty, &$div, &$mod) {
  return function($n) use (&$push, &$if, &$isLessOrEqual, &$decrement, &$ten, &$empty, &$div, &$mod, &$f) {
    return $push(
      $if($isLessOrEqual($n)($decrement($ten)))(
        $empty
      )(
        function($x) use(&$div, &$ten, &$f, &$n) {
          return $f($div($n)($ten))($x);
        }
      )
    )($mod($n)($ten));
  };
});
?>

File: ruby/lambda.rb
def to_integer(proc)
  proc[-> n { n + 1 }][0]
end

def to_boolean(proc)
  IF[proc][true][false]
end

def to_array(proc)
  array = []

  until to_boolean(IS_EMPTY[proc])
    array.push(FIRST[proc])
    proc = REST[proc]
  end

  array
end

def to_char(c)
  '0123456789BFiuz'.slice(to_integer(c))
end

def to_string(s)
  to_array(s).map { |c| to_char(c) }.join
end

ZERO  = -> p { -> x {       x    } }
ONE   = -> p { -> x {     p[x]   } }
TWO   = -> p { -> x {   p[p[x]]  } }
THREE = -> p { -> x { p[p[p[x]]] } }

FIVE    = -> p { -> x { p[p[p[p[p[x]]]]] } }
TEN     = -> p { -> x { p[p[p[p[p[p[p[p[p[p[x]]]]]]]]]] } }
FIFTEEN = -> p { -> x { p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[x]]]]]]]]]]]]]]] } }
HUNDRED = -> p { -> x { p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[p[x]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]] } }

TRUE  = -> x { -> y { x } }
FALSE = -> x { -> y { y } }

IF = -> b { b }

IS_ZERO = -> n { n[-> x { FALSE }][TRUE] }

INCREMENT = -> n { -> p { -> x { p[n[p][x]] } } }
DECREMENT = -> n { -> f { -> x {
  n[-> g { -> h { h[g[f]] } }][-> u { x }][-> u { u }]
} } }

ADD      = -> m { -> n { n[INCREMENT][m] } }
SUBTRACT = -> m { -> n { n[DECREMENT][m] } }
MULTIPLY = -> m { -> n { n[ADD[m]][ZERO] } }
POWER    = -> m { -> n { n[MULTIPLY[m]][ONE] } }

IS_LESS_OR_EQUAL = -> m { -> n { IS_ZERO[SUBTRACT[m][n]] } }

Z = -> f { -> x { f[-> y { x[x][y] }] }[-> x { f[-> y { x[x][y] }] }] }

MOD = Z[-> f { -> m { -> n {
  IF[IS_LESS_OR_EQUAL[n][m]][
    -> x { f[SUBTRACT[m][n]][n][x] }
  ][
    m
  ]
} } }]

DIV = Z[-> f { -> m { -> n {
  IF[IS_LESS_OR_EQUAL[n][m]][
    -> x { INCREMENT[f[SUBTRACT[m][n]][n]][x] }
  ][
    ZERO
  ]
} } }]

PAIR  = -> x { -> y { -> f { f[x][y] } } }
LEFT  = -> p { p[-> x { -> y { x } } ] }
RIGHT = -> p { p[-> x { -> y { y } } ] }

EMPTY     = PAIR[TRUE][TRUE]
UNSHIFT   = -> l { -> x {
              PAIR[FALSE][PAIR[x][l]]
            } }
IS_EMPTY  = LEFT
FIRST     = -> l { LEFT[RIGHT[l]] }
REST      = -> l { RIGHT[RIGHT[l]] }

RANGE = Z[-> f { -> m { -> n {
  IF[IS_LESS_OR_EQUAL[m][n]][
    -> x { UNSHIFT[f[INCREMENT[m]][n]][m][x] }
  ][
    EMPTY
  ]
} } }]

FOLD = Z[-> f { -> l { -> x { -> g {
  IF[IS_EMPTY[l]][
    x
  ][
    -> y { g[f[REST[l]][x][g]][FIRST[l]][y] }
  ]
} } } }]

MAP = -> k { -> f {
  FOLD[k][EMPTY][
    -> l { -> x { UNSHIFT[l][f[x]] } }
  ]
} }

PUSH = -> l { -> x { FOLD[l][UNSHIFT[EMPTY][x]][UNSHIFT] } }

B   = TEN
F   = INCREMENT[B]
I   = INCREMENT[F]
U   = INCREMENT[I]
ZED = INCREMENT[U]

FIZZ     = UNSHIFT[UNSHIFT[UNSHIFT[UNSHIFT[EMPTY][ZED]][ZED]][I]][F]
BUZZ     = UNSHIFT[UNSHIFT[UNSHIFT[UNSHIFT[EMPTY][ZED]][ZED]][U]][B]
FIZZBUZZ = UNSHIFT[UNSHIFT[UNSHIFT[UNSHIFT[BUZZ][ZED]][ZED]][I]][F]

TO_DIGITS = Z[-> f { -> n {
  PUSH[
    IF[IS_LESS_OR_EQUAL[n][DECREMENT[TEN]]][
      EMPTY
    ][
      -> x { f[DIV[n][TEN]][x] }
    ]
  ][MOD[n][TEN]]
} }]
